### DISCUSSION TITLE
Eccentric finalization in C# interactive

### DISCUSSION THREAD

**Discussion by deep-outcome** [dotnet/roslyn#56547] [Q&A] [answered]
**Version Used**: 

Microsoft (R) Visual C# Interactive Compiler version 3.10.0-4.21318.11 ()

### Describe the bug
It seems as if GC has some trouble with 1st execution of finalizers.

```csharp
> class A
. {
.     ~A () => Write ('A');
. }
> class B : A, IDisposable
. {
.     public void Dispose () => GC.SuppressFinalize (this);
. };
> new B (); GC.Collect ();  // No 'A' written.
> new B (); GC.Collect (); 
A
> new B (); GC.Collect (); 
A
> new B (); GC.Collect (); 
A
> new B ().Dispose(); GC.Collect (); // 'A' written.
A
> new B ().Dispose(); GC.Collect ();
> new B ().Dispose(); GC.Collect ();
> new B ().Dispose(); GC.Collect ();
```
**Steps to Reproduce**:

1. Try code from sample above.

**Expected Behavior**:
GC/Finalizer behavior is consistent.

**Actual Behavior**:
There is some unobvious behavior.

**Comment by CyrusNajmabadi** [comment]
I see four 'A's written for four non-disposed B's created.  This seems appropriate to me.  I'm betting if you have it print out something other than 'A' (like something you pass the constructor), you'll see all the non-disposed B's print out their unique strings, and you'll see the disposed ones do not.

**Comment by PathogenDavid** [comment, marked as answer]
As Cyrus noted: Your objects are being collected appropriately, just not when you expect. Finalizers do not run synchronously with collection, and the exact time they are called and which thread they are called from is unspecified.

From [the documentation](https://docs.microsoft.com/en-us/dotnet/api/system.object.finalize?view=net-5.0#how-finalization-works):

> Finalize operations have the following limitations:
> * The exact time when the finalizer executes is undefined. To ensure deterministic release of resources for instances of your class, implement a `Close` method or provide a [IDisposable.Dispose](https://docs.microsoft.com/en-us/dotnet/api/system.idisposable.dispose) implementation.
> * [...]
> * The thread on which the finalizer runs is unspecified.

Your situation is further complicated by the fact that the garbage collector considers your `B` instances to still be reachable* when you're calling `GC.Collect`, so it's not actually even being collected when you call `GC.Collect`. (If you submit `new B()` and `GC.Collect()` separately you'll find it's finalized seemingly instantly.)

*The JIT does not always tell the GC that references for the current stack frame are no longer in use, and never does for debug builds. ([Ref](https://github.com/Maoni0/mem-doc/blob/725e4cb78db706bb13346b3f6ef4a62228315de1/doc/.NETMemoryPerformanceAnalysis.md#2-user-roots))

**Reply by deep-outcome** [reply]
My fault. I did not know (or consider) _`new B()`_ reachability. 

```csharp
> new B ();
> GC.Collect ();
A
> new B ();
> GC.Collect ();
A
> new B ().Dispose();
> GC.Collect ();
> new B ().Dispose();
> GC.Collect ();
>
```
